libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFOAuthCoordinator.m

```diff
@@ -519,86 +535,76 @@ - (void)beginUserAgentFlow {
-                        NSString *escapedString = [approvalUrl stringByURLEncoding];
-                        NSString* approvalUrl = [NSString stringWithFormat:@"%@/secur/frontdoor.jsp?sid=%@&retURL=%@", baseUrlString, self.credentials.accessToken, escapedString];
-                        [self doLoadURL:approvalUrl withCookie:YES];
-                        swapOK = YES;
-                        self.credentials.jwt = nil;
-                    }
-                }
-                if (!swapOK) {
-                    NSError *error = [[self class] errorWithType:kSFOAuthErrorTypeJWTLaunchFailed description:@"The breeze link failed to launch."];
-                    [self notifyDelegateOfFailure:error authInfo:self.authInfo];
-                    self.credentials.jwt = nil;
-                }
-            }
-            else {
-                [self log:SFLogLevelError msg:[NSString stringWithFormat:@"Fail to swap JWT for access token: %@", [error localizedDescription]]];
+    NSAssert(self.credentials.jwt.length > 0, @"JWT token should be present at this point.");
+    
+    [self swapJWTWithCompletionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
+        if (error != nil) {
+            [self log:SFLogLevelError format:@"Fail to swap JWT for access token: %@", error.localizedDescription];
+            [self notifyDelegateOfFailure:error authInfo:self.authInfo];
+            return;
+        }
+        
+        self.credentials.jwt = nil;
+        NSError *jsonError = nil;
+        id json = nil;
+        
+        json = [NSJSONSerialization JSONObjectWithData:data options:0 error:&jsonError];
+        if (jsonError != nil) {
+            NSError *error = [[self class] errorWithType:kSFOAuthErrorTypeJWTLaunchFailed
+                                                 description:@"Error parsing JWT token exchange response."
+                                             underlyingError:jsonError];
                 [self notifyDelegateOfFailure:error authInfo:self.authInfo];
-            }
-        }];
-    }
-    else {
-        [self doLoadURL:approvalUrl withCookie:NO];
-    }
+            return;
+        }
+        
+        if (![json isKindOfClass:[NSDictionary class]]) {
+            NSString *errorDesc = [NSString stringWithFormat:@"Expected NSDictionary for JWT token response, received %@ instance.", NSStringFromClass([json class])];
+                NSError *error = [[self class] errorWithType:kSFOAuthErrorTypeJWTLaunchFailed
+                                                 description:errorDesc];
+            [self notifyDelegateOfFailure:error authInfo:self.authInfo];
+            return;
+        }
+        
+        NSDictionary *dict = (NSDictionary *)json;
+        if (nil != dict[kSFOAuthError]) {
+            NSError *error = [[self class] errorWithType:dict[kSFOAuthError] description:dict[kSFOAuthErrorDescription]];
```

Comment:
This is a change from the previous implementation.  The old implementation masked any error returned from the service here, and this implementation embraces it, as the other flows do.